fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -1,155 +1,175 @@
-) -> anyhow::Result<()> {
-    let mut decision_receivers = vec![];
-    for ordered_item in signed_block.block.items.into_iter() {
-        let (decision_sender, decision_receiver) = oneshot::channel();
+            tokio::spawn(async move {
+                let mut decision_receivers = vec![];
 
-        ordered_item_sender
-            .send((ordered_item, signed_block.block.index, decision_sender))
-            .await?;
+                for ordered_item in signed_block.block.items.into_iter() {
+                    let (decision_sender, decision_receiver) = oneshot::channel();
 
-        decision_receivers.push(decision_receiver);
-    }
+                    ordered_item_sender
+                        .send(Some((ordered_item, decision_sender)))
+                        .await
+                        .ok();
 
-    for decision_receiver in decision_receivers {
-        // The threshold signed blocks items have to be accepted by Fedimint Consensus.
-        assert!(decision_receiver.await? == Decision::Accept);
-    }
+                    decision_receivers.push(decision_receiver);
+                }
 
-    Ok(())
-}
+                // signal that the session is complete
+                ordered_item_sender.send(None).await.ok();
 
-#[allow(clippy::too_many_arguments)]
-pub async fn run(
-    keychain: Keychain,
-    db: Database,
-    mut index: u64,
-    mempool_item_receiver: async_channel::Receiver<Vec<u8>>,
-    incoming_message_receiver: async_channel::Receiver<(Message, PeerId)>,
-    outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
-    ordered_item_sender: mpsc::Sender<(OrderedItem, u64, oneshot::Sender<Decision>)>,
-    clean_shutdown_receiver: watch::Receiver<Option<(u64, Duration)>>,
-) -> Shutdown {
-    let (network_data_sender, network_data_receiver) = async_channel::bounded(256);
-    let (signed_block_sender, signed_block_receiver) = async_channel::bounded(16);
-
-    let relay_handle = tokio::spawn(relay_messages(
-        db.clone(),
-        incoming_message_receiver,
-        outgoing_message_sender.clone(),
-        network_data_sender,
-        signed_block_sender,
-    ));
-
-    loop {
-        if let Some(signed_block) = db::load_block(&db, index).await {
-            tracing::info!("Loaded block with index {}", index);
-            if process_signed_block(signed_block, ordered_item_sender.clone())
-                .await
-                .is_err()
-            {
-                relay_handle.abort();
-                relay_handle.await.ok();
-
-                return Shutdown::MidSession(index);
-            }
+                for decision_receiver in decision_receivers {
+                    // The items in a threshold signed block have to be accepted
+                    if let Ok(decision) = decision_receiver.await {
+                        assert!(decision == Decision::Accept)
+                    }
+                }
+            });
```

Comment:
How is this shut down when fedimint is shut down ([`TaskGroup`s](https://docs.fedimint.org/fedimint_core/task/struct.TaskGroup.html) would help here)?